Upgrade to functional views

JannikStreek opened the pull request:
## Further Notes

closes #357 

- [Additional context information](https://dev.to/byronsalty/upgrading-to-phoenix-17-3l21)
- https://elixircasts.io/upgrading-to-phoenix-1.7
- https://elixirforum.com/t/notes-on-1-6-to-1-7-upgrade-with-verified-routes-and-phoenix-view-removal/54759
- Live view docs for forms: https://hexdocs.pm/phoenix_live_view/form-bindings.html

## To be migrated
- [x] landing page / static page
- [x] layouts
- [x] admin
- [x] errors
- [x] live/brainstorming
- [x] live/ideas
- [x] Fix label display
- [x] fix html export
- [x] Fix sigil lib/mindwendel_web/templates/error/error_page.html.heex

The error view was removed so only the template is now left, see `config.exs`:
```
# Configures the endpoint
config :mindwendel, MindwendelWeb.Endpoint,
  render_errors: [
    view: MindwendelWeb.ErrorView,
    accepts: ~w(html json),
    layout: false
  ],
```

## After Merge Checklist

- [ ] Fix remaining issues: A label in use can't be deleted => error message isn't shown

JannikStreek committed:
started changing static page to function component

JannikStreek committed:
save

JannikStreek committed:
move files

JannikStreek committed:
one page works

JannikStreek committed:
fix routes and added form back in

JannikStreek committed:
refactor util function call

JannikStreek committed:
remove layout view

JannikStreek committed:
landing page view works

JannikStreek committed:
update brainstorming controller

JannikStreek committed:
added back removed root content

JannikStreek committed:
add back plug

JannikStreek commented on lib/mindwendel_web.ex:
```diff
@@ -17,13 +17,68 @@ defmodule MindwendelWeb do
+  def html do
+    quote do
+      # this is temporary https://hexdocs.pm/phoenix_view/Phoenix.View.html#module-replaced-by-phoenix-component
+      # import Phoenix.View
```
can be removed if everything works out later - currently not used anyway

JannikStreek committed:
fix live view module loading

JannikStreek committed:
include live helpers

JannikStreek committed:
fix link in edit

JannikStreek committed:
wip admin/brainstorming_live/edit.html.heex

PragTob commented on lib/mindwendel_web/controllers/static_page_html/home.html.heex:
```diff
@@ -1,10 +1,5 @@
-      src={Routes.static_path(@conn, "/images/mindwendel_logo_white.svg")}
-      height="40px"
-      class="float-md-start me-3"
-    />
```
que? (why did you remove the image without replacement?)

JannikStreek replied:
needs to be added back in. everything content wise that was removed was just temporary so I could temporarily remove one more error message. in the end all content should be in again. Guess it was during my testing to get the new route feature working.

PragTob commented on lib/mindwendel_web/live/idea_live/index_component.html.heex:
```diff
@@ -3,108 +3,16 @@
         data-position={idea.position_order}
       >
         <div class="card-body-mindwendel-idea">
-          <%= if @current_user.id in [idea.user_id | @brainstorming.moderating_users |> Enum.map(& &1.id)] do %>
```
a bit confused how we can delete all that? :thinking:

PragTob replied:
we definitely can't but it dawns on me that maybe it was done because of the lanes but I'm not sure

PragTob committed:
Restore basic Admin edit functionality

HTML report may remain broken, but it renders and the sockets
also work again due to me patching in the meta tage again.

Hope I didn't break anything, pls review ;)

PragTob committed:
Merge branch 'master' into upgrade-to-functional-views

PragTob commented on lib/mindwendel_web/templates/layout/root.html.heex:
```diff
@@ -7,17 +7,12 @@
       content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0"
     />
-    <%= csrf_meta_tag() %>
+    <meta name="csrf-token" content={get_csrf_token()} />
```
added this one back here similar to the static.html.eex for now, I guess you shouldn't need it in a liveview but app.js was crashing on:

```
let csrfToken = document.querySelector("meta[name='csrf-token']").getAttribute("content")
```

so none of the "live" features worked so I added it back in for now

PragTob committed:
leave notes about HTML export

PragTob commented on lib/mindwendel_web/live/brainstorming_live/show.html.heex:
```diff
@@ -21,100 +21,17 @@
-      brainstorming: @brainstorming,
-      id: @brainstorming.id,
-      current_user: @current_user
-    ) %>
```
Jannik so: Lol let's see how long Tobi takes until he finds out I removed the ideas from rendering in a brainstorming :angel:

JannikStreek replied:
😂 I sometimes start with just removing stuff and introducing it back in one by one. the problem currently is that there are so many errors that you would need to remove basically everything 😂

PragTob committed:
make brainstorming#show render

PragTob committed:
Introduce rendering ideas, random path fixes

I'm not 100% sure if taking out ideas was a mistake or if it was
meant as such so I'm not pushing forward with making more of them
visible as even more had been commented out there.

Removed more warnings, however some I was unsure about.

Left some TODO comments on places in the code where I gained
knowledge/know of work to do.

PragTob committed:
Remove patch to users path which seemed accidental

PragTob committed:
make verified routes helper available in conn case tests

Test doesn't work yet because it tests on a whole view which
is borked yet but still an increment :)

JannikStreek commented on lib/mindwendel_web/templates/error/error_page.html.heex:
```diff
@@ -8,6 +8,7 @@
       content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0"
     />
     <title>mindwendel</title>
+    <%!-- TODO: ~p isn't known/used here and so also fails, but I'm also not familiar with this page --%>
```
I think it's still being called from a view and views don't have sigil available? Best way would be to refactor errors into components I think so we can remove all view files.

PragTob replied:
Agreed. I'm not sure though - I think views should also have ~p available, it's probably just the matter of a missing import.

PragTob replied:
@JannikStreek yup https://hexdocs.pm/phoenix/Phoenix.VerifiedRoutes.html#module-options

`use Phoenix.VerifiedRoutes` should solve the problem

PragTob replied:
We should still move it to components overall but maybe that's a stepping stone :)

JannikStreek committed:
add buttons to brainstorming page back in

JannikStreek committed:
add back modal new idea

JannikStreek committed:
add idea modal working

JannikStreek committed:
add back logo

JannikStreek committed:
add back idea cards

JannikStreek committed:
add back uuid

JannikStreek committed:
idea index working again

JannikStreek commented on lib/mindwendel_web/live/idea_live/index_component.html.heex:
```diff
@@ -37,27 +48,26 @@
-              <%= text_to_html(idea.body) %>
-            <% end %>
+            <.link href={idea.link.url}>
+              <%= raw idea.body %>
```
`text_to_html ` is no longer available

PragTob replied:
@JannikStreek but this right now allows for XSS, and also doesn't insert the line breaks as text_to_html did. txt_to_html has escaped HTML elements by default.

I guess there is some other function now that does this, but I think I'd rather than using `raw` have the text be escaped.

PragTob replied:
I think the function got moved the HTML helpers but there should be a function component with its functionality or we can write one:

https://github.com/phoenixframework/phoenix_html/blob/main/CHANGELOG.md#v400-2023-12-19

> This version removes deprecated functionality and moved all HTML helpers to a separate library. HTML Helpers are no longer used in new apps from Phoenix v1.7, instead it relies on function components from Phoenix.LiveView. Older applications who wish to maintain compatibility, add {:phoenix_html_helpers, "~> 1.0"} to your mix.exs and then replace use Phoenix.HTML in your applications by

JannikStreek replied:
ah good to know. I just threw out views completely and it still works 😅 , so close to finishing this :)

JannikStreek committed:
share modal working

JannikStreek committed:
improve modal

JannikStreek committed:
styled inputs

JannikStreek committed:
upgrade live view and remove live dashboard as it is in conflict with the update

JannikStreek committed:
errors now working for idea fornm

JannikStreek committed:
remove view dependency as no longer needed

JannikStreek committed:
fix warnings

JannikStreek committed:
fix routes in test

JannikStreek committed:
fix test

JannikStreek committed:
label editing in admin works again

JannikStreek committed:
fix tests, one test is commented and will be extracted to a new issue

JannikStreek committed:
revert live view upgrade and fix confirm warning

JannikStreek committed:
update locals

JannikStreek committed:
fix design of labels in admin

JannikStreek committed:
fix html export

JannikStreek committed:
missed files

JannikStreek committed:
remove todo

JannikStreek committed:
add error files

JannikStreek committed:
remove todo

JannikStreek committed:
update package lock

JannikStreek committed:
update language files

JannikStreek committed:
Merge branch 'master' into upgrade-to-functional-views

JannikStreek committed:
fix reload when labels error

JannikStreek committed:
Merge remote-tracking branch 'origin/upgrade-to-functional-views' into upgrade-to-functional-views
